Fix firewall rules

SantiagoTorres opened the pull request:
Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

**Fixes issue #**: 

None, this issue was brought up from conversations with potential integrators

**Description of the changes being introduced by the pull request**:

The current behavior of the artifact rules throw exceptions when an element in the *filtered* queue doesn't match the target. For example, if a file `foo` is being checked for a `MODIFY` rule then a `VerificationError` will be raised if `foo` is indeed not modified. This is not how firewalls behave.

Instead, `foo` should just remain in the queue and expected to be matched by other rules later down the line. The only rule that should throw an error is a `DISALLOW` rule when artifacts are matched.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

This is still somewhat of a WIP branch, as I'd like to have some feedback from @vladimir-v-diaz and @lukpueh (if possible)

SantiagoTorres committed:
ENH: verifylib: fix verification behavior

The verification behavior didn't behave as a firewall, as described in
the documentation. Instead, it used to throw a verification exception if
sctrict matching didn't happen.

Update the behavior of the in_toto_verify routines, as well as the
specific subroutines to match the expected behavior.

SantiagoTorres committed:
FIX:test_verifylib: update test to match verifylib

Verifylib was changed in the parent commit. Update the tests to match
this behavior by, mostly, removing the expected exception raises and
adding "DISALLOW *" rules where necessary.

SantiagoTorres committed:
FIX: CHANGELOG: address changelog review comments

- The pinned version number was wrong
- Added mention to the obsolete settings removal.

SantiagoTorres commented:
FYI: these failures appear to be due to the set operator using a different `PYTHONHASHSEED`

vladimir-v-diaz commented:
@JustinCappos requested that @awwad review more in-toto pull requests, so I'll add Sebastien to the list of reviewers.  For now, I need to focus on fixing Windows-related issues for the TUF project.

SantiagoTorres commented:
> @JustinCappos requested that @awwad review more in-toto pull requests, so I'll add Sebastien to the list of reviewers. For now, I need to focus on fixing Windows-related issues for the TUF project.

Noted, and thanks!

@awwad welcome aboard!

SantiagoTorres committed:
FIX:test_verifylib: sort artifact queues

During the test of the modify rules, it is possible that the returned
queues are in different order than they came in (this is due to the
converstion to a set and back).

Sort the queues before comparing the two queues to ensure the contents
are compared, and not their order. Likewise, declare the elements in artifact
queues alfabetically to make sure the comparison stands.

awwad commented:
So if I understand correctly, with this PR merged, if you have some rule set wherein rule 1 reads: `MATCH arty WITH PRODUCTS FROM step1`,  rules 2 through 5 have nothing to do with arty, and rule 6 is `DISALLOW *`, then the error you get when arty exists and has a different hash than that listed for arty in the step1.link products is an error about arty being disallowed by rule 6, right?

awwad commented:
Also, with this construction (rule verifier functions adjust the input to remove what they've matched), none of the hash mismatch info gets reported anywhere -- neither trickles up in an exception that's caught somewhere nor logged. Is that okay in this context? Is that going to be regrettable eventually?

In any case, the PR seems sane.

awwad committed:
Minor: typo fix in test func name

test_fail_hash_not_eual -> test_fail_hash_not_equal

Signed-off-by: Sebastien Awwad <sebastien.awwad@gmail.com>

SantiagoTorres commented:
wow, sorry for some reason I missed this PR review. 

> So if I understand correctly, with this PR merged, if you have some rule set wherein rule 1 reads: MATCH arty WITH PRODUCTS FROM step1, rules 2 through 5 have nothing to do with arty, and rule 6 is DISALLOW *, then the error you get when arty exists and has a different hash than that listed for arty in the step1.link products is an error about arty being disallowed by rule 6, right?

Yes, this is pretty much it. We're trying to have DISALLOW be the only rule that fails, where the others match and process. That's the behavior we had described in the specification.

> Also, with this construction (rule verifier functions adjust the input to remove what they've matched), none of the hash mismatch info gets reported anywhere -- neither trickles up in an exception that's caught somewhere nor logged. Is that okay in this context? Is that going to be regrettable eventually?

We were worried about this too, but it's a shared issue on firewall-like semantics. I don't think we can make much about it (yet).

> In any case, the PR seems sane.

Thanks a lot! Merging :)
